src/app-layer-expectation.c

```diff
@@ -0,0 +1,287 @@
+ *
+ * Create a expectation from an existing Flow. Currently, only Flow between
+ * the two original IP addresses are supported.
+ *
+ * \param f a pointer to the original Flow
+ * \param direction the direction of the expectation flow with respect to original flow
+ * \param src source port of the expected flow, use 0 for any
+ * \param dst destination port of the expected flow, use 0 for any
+ * \param alproto the protocol that need to be set on the expected flow
+ * \param data pointer to data that will be attached to the expected flow
+ *
+ * \return -1 if error
+ * \return 0 if success
+ */
+
+int AppLayerExpectationCreate(Flow *f, int direction, Port src, Port dst, AppProto alproto, void *data)
+{
+    Expectation *exp = SCCalloc(1, sizeof(*exp));
+    Expectation *iexp = NULL;
+    IPPair *ipp;
+    Address ip_src, ip_dst;
+
+    if (exp == NULL)
+        return -1;
+
+    exp->sp = src;
+    exp->dp = dst;
+    exp->alproto = alproto;
+    exp->ts = f->lastts;
+    exp->data = data;
+    exp->direction = direction;
+
+    if (GetFlowAddresses(f, &ip_src, &ip_dst) == -1)
+        goto error;
+    if (direction & STREAM_TOSERVER) {
+        ipp = IPPairGetIPPairFromHash(&ip_src, &ip_dst);
+    } else {
+        ipp = IPPairGetIPPairFromHash(&ip_dst, &ip_src);
+    }
+    if (ipp == NULL)
+        goto error;
+
+    iexp = IPPairGetStorageById(ipp, expectation_id);
+    exp->next = iexp;
+    IPPairSetStorageById(ipp, expectation_id, exp);
+
+    SC_ATOMIC_ADD(expectation_count, 1);
+    IPPairUnlock(ipp);
+    return 0;
+
+error:
+    if (exp != NULL)
+        SCFree(exp);
+    return -1;
+}
+
+static Expectation *AppLayerExpectationGet(Flow *f, int direction, IPPair **ipp)
+{
+    Address ip_src, ip_dst;
+    if (GetFlowAddresses(f, &ip_src, &ip_dst) == -1)
+        return NULL;
+    if (direction & STREAM_TOSERVER) {
+        *ipp = IPPairLookupIPPairFromHash(&ip_src, &ip_dst);
+    } else {
+        *ipp = IPPairLookupIPPairFromHash(&ip_dst, &ip_src);
+    }
+    if (*ipp == NULL) {
+        return NULL;
+    }
+
+    return IPPairGetStorageById(*ipp, expectation_id);
+}
+
+/**
+ * Return Flow storage identifier corresponding to expectation data
+ *
+ * \return expectation data identifier
+ */
+int AppLayerExpectationGetDataId(void)
+{
+    return expectation_data_id;
+}
+
+static Expectation * RemoveExpectationAndGetNext(IPPair *ipp,
+                                Expectation *pexp, Expectation *exp,
+                                Expectation *lexp)
+{
+    SC_ATOMIC_SUB(expectation_count, 1);
+    if (pexp == NULL) {
+        IPPairSetStorageById(ipp, expectation_id, lexp);
+    } else {
+        pexp->next = lexp;
+    }
+    if (exp->data)
+        SCFree(exp->data);
+    SCFree(exp);
+    return lexp;
+}
+
+/**
+ * Function doing a lookup in expectation list
+ *
+ * \return an AppProto value if found
+ * \return ALPROTO_UNKNOWN if not found
+ */
+AppProto AppLayerExpectationLookup(Flow *f, int direction)
```

Comment:
when I see 'Lookup' in the func name, I'm not expecting things like FlowSetStorageById to happen. Should this be called differently? E.g. ...Handle()?

Reply:
yes make sense

Reply:
in general I think this func needs an explanation of what it does. It *looks* like it transfers and expectation from ippair to flow and then does some house keeping. It should explain that in comment

Reply:
Adding comment about the implemented process in the doxygen comment.